File: genql/compilers/policy_compiler.py
"""
genql.compilers.policy_compiler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compiles POLICY intents into a declarative policy DSL.

Supported body keys
-------------------
subject : str
    The actor or entity the rule applies to (e.g. ``"user"``, ``"service"``).
action : str
    The operation being governed (e.g. ``"read"``, ``"write"``, ``"delete"``).
resource : str
    The resource being protected (e.g. ``"document"``, ``"account"``).
effect : str
    ``"allow"`` or ``"deny"``.
conditions : list[str]  (optional)
    Human-readable predicate strings that must hold for the effect to apply.
priority : int  (optional, default 0)
    Higher priority rules override lower ones on conflict.
"""

from __future__ import annotations

from typing import List

from genql.compilers.base import Compiler
from genql.intent import Intent, IntentKind


class PolicyCompiler(Compiler):
    """Translates :attr:`IntentKind.POLICY` intents to a policy DSL."""

    @property
    def target_language(self) -> str:
        return "policy_dsl"

    @property
    def supported_kinds(self) -> List[IntentKind]:
        return [IntentKind.POLICY]

    def compile(self, intent: Intent) -> str:
        body = intent.body
        subject = body.get("subject", "any")
        action = body.get("action", "any")
        resource = body.get("resource", "any")
        effect = str(body.get("effect", "deny")).upper()
        conditions: List[str] = list(body.get("conditions") or [])
        priority = int(body.get("priority", 0))

        lines = [
            f"POLICY {intent.name!r} [priority={priority}]:",
            f"  SUBJECT  {subject}",
            f"  ACTION   {action}",
            f"  RESOURCE {resource}",
            f"  EFFECT   {effect}",
        ]
        if conditions:
            lines.append("  WHEN")
            for cond in conditions:
                lines.append(f"    AND {cond}")

        return "\n".join(lines)

File: genql_ruby/lib/genql/compilers/policy_compiler.rb
# genql/compilers/policy_compiler
#
# Compiles POLICY intents into a declarative policy DSL.
#
# Supported body keys
# -------------------
# subject    : String       — the actor or entity the rule applies to
# action     : String       — the operation being governed
# resource   : String       — the resource being protected
# effect     : String       — "allow" or "deny"
# conditions : Array<String> — predicates that must hold (optional)
# priority   : Integer      — higher priority overrides lower on conflict (default 0)

require_relative "base"
require_relative "../intent"

module Genql
  module Compilers
    class PolicyCompiler < Compiler
      def target_language
        "policy_dsl"
      end

      def supported_kinds
        [IntentKind::POLICY]
      end

      def compile(intent)
        body       = intent.body
        subject    = body_get(body, "subject")   || "any"
        action     = body_get(body, "action")    || "any"
        resource   = body_get(body, "resource")  || "any"
        effect     = (body_get(body, "effect")   || "deny").to_s.upcase
        conditions = Array(body_get(body, "conditions"))
        priority   = (body_get(body, "priority") || 0).to_i

        lines = [
          "POLICY '#{intent.name}' [priority=#{priority}]:",
          "  SUBJECT  #{subject}",
          "  ACTION   #{action}",
          "  RESOURCE #{resource}",
          "  EFFECT   #{effect}"
        ]

        if conditions.any?
          lines << "  WHEN"
          conditions.each { |cond| lines << "    AND #{cond}" }
        end

        lines.join("\n")
      end

      private

      def body_get(body, key)
        body[key.to_sym] || body[key.to_s]
      end
    end
  end
end
